queryscript/src/compile/generics.rs

```diff
@@ -180,15 +190,94 @@ impl Generic for SumGeneric {
     }
 
     fn unify(&self, other: &MType) -> Result<()> {
-        // This is a bit of an approximate implementation, since it only works if the inner
-        // type is known.
-        if self.0.is_known()? {
-            let final_type =
-                MType::from_runtime_type(&self.to_runtime_type().context(RuntimeSnafu {
-                    loc: ErrorLocation::Unknown,
-                })?)?;
-            other.unify(&final_type)?;
-        }
+        let other = other.clone();
+        let arg = self.0.clone();
+        arg.constrain(move |arg: Ref<MType>| {
+            let arg = arg.read()?.to_runtime_type().context(RuntimeSnafu {
+                loc: ErrorLocation::Unknown,
+            })?;
+            let final_type = MType::from_runtime_type(&arg)?;
+            other.unify(&final_type)
+        })?;
+        Ok(())
+    }
+}
+
+pub struct CoerceGeneric {
+    op: CoerceOp,
+    args: Vec<CRef<MType>>,
+}
+
+impl CoerceGeneric {
+    pub fn new(op: CoerceOp, args: Vec<CRef<MType>>) -> CoerceGeneric {
+        CoerceGeneric { op, args }
+    }
+
+    fn static_name() -> &'static Ident {
+        &COERCE_GENERIC_NAME
+    }
+}
+
+fn coerce_list(op: CoerceOp, args: Vec<CRef<MType>>) -> runtime::error::Result<types::Type> {
+    if args.len() == 0 {
+        return Ok(types::Type::Atom(AtomicType::Null));
+    }
+    let runtime_args = args
+        .iter()
+        .map(|arg| Ok(arg.must()?.read()?.to_runtime_type()?))
+        .collect::<runtime::error::Result<Vec<_>>>()?;
+
+    let mut ret = runtime_args[0].clone();
+    for arg in &runtime_args[1..] {
+        ret = coerce_types(&ret, &op, &arg).ok_or_else(|| runtime::error::RuntimeError::new(""))?;
```

Comment:
Is this your replacement to `MType::coerce`? If so, why are we returning an empty `RuntimeError` here instead of the nice `CompileError::coercion` we threw before?

Reply:
Good catch, will fix ++